### ISSUE TITLE
Race condition on service startup (DB migration, route setup)

### ISSUE THREAD

**Issue by eliasp** [rseichter/automx2#29]
When starting automx2 and immediately initializing the DB using the `/initdb` endpoint, in most cases I was able to reproduce a race condition, where the DB migration didn't set up the schema yet while the service is already accepting requests on `/initdb`, resulting in an exception being thrown:

```
flask[294436]: (sqlite3.OperationalError) no such table: provider                                                                                                                                                     
flask[294436]: [SQL: SELECT count(*) AS count_1                                                                                                                                                                       
flask[294436]: FROM (SELECT provider.id AS provider_id, provider.name AS provider_name, provider.short_name AS provider_short_name                                                                                    
flask[294436]: FROM provider) AS anon_1]
flask[294436]: (Background on this error at: https://sqlalche.me/e/20/e3q8)
```

I tried fixing this by pulling the [`Migrate()`](https://github.com/rseichter/automx2/blob/36098d6/automx2/server.py#L61) call in front of all [`app.add_url_rule()`](https://github.com/rseichter/automx2/blob/36098d6/automx2/server.py#L52-L57) statements, but this didn't fix the issue - it seems there's some async operation going on in the background finalizing the DB migration.

Furthermore, there also doesn't seem to be a way (based on my limited Flask knowledge/research results) to initialize a Flask app in a way, where it signals service readiness (e.g. via [`sd_notify()`](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#)) only once all routes are added and the DB is initialized, which would basically eliminate this problem altogether, since systemd's `ExecStartPost=` would only be executed once the MainPID is ready.

For now, this is solved using an ugly `sleep 3` before the 1st call to the API and then a while loop which retries a curl statement to `/initdb` a few times until it succeeds, but this feels more like a band-aid which is prone to break and introduces an unneeded startup performance penalty than a proper solution.

**Comment by rseichter** [comment]
I am trying to imagine your use case. Can you provide more details?

From a design perspective, the initial DB setup is expected to run only once. Before automx2 can deliver meaningful responses, the database then needs to be populated. 

Only after these steps are completed should the proxy server forward requests to automx2, and that proxy is also the natural point to limit access. The route `/initdb/` should be blocked there.

**Comment by eliasp** [comment]
This originates from the setup provided by NixOS `service.automx2` module.
- [current implementation](https://github.com/NixOS/nixpkgs/blob/a68c330/nixos/modules/services/mail/automx2.nix)
- [improvements that I'm working on right now (WIP)](https://github.com/eliasp/nixpkgs/blob/automx2-improvements/nixos/modules/services/mail/automx2.nix)

As of now, it only supports in-memory SQLite as backend, since the current implementation within the boundary conditions implied by NixOS don't allow for secure credential handling for other DBs right now - I might follow up with another issue in case this needs upstream work.

On each service startup, an `ExecStartPost=` uses a `curl` call sending JSON to `/initdb` to initialize the DB according to the service's configuration, so the service is completely stateless as of now.

The problem was: I believed, the DB migration wouldn't be a synchronous operation, but would instead happen in the background.

The actual problem was: my winter holidays brain made assumptions about `ExecStart=` fully starting `automx2` before continuing with `ExecStartPost=`, while all it does is launching the process, but since it's not a `Type=notify` process, it has zero knowledge about the service's internal state, so often it tried to talk to `/initdb` even before the listener was up or the DB was initialized.

The diff below solves this problem by adding support for systemd's [`sd_notify`](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html) protocol which allows a service to inform systemd about its state/readiness.

For now, I'll use this downstream, but feel free to re-use it in whatever way you want or just add it under your name to prevent having to care about your contribution policy.

```diff
From 32a6e038225716fd30c939d90973ef7e8aa7f5bd Mon Sep 17 00:00:00 2001
From: Elias Probst <mail@eliasprobst.eu>
Date: Wed, 1 Jan 2025 16:30:22 +0100
Subject: [PATCH] Add sdnotify support

Signed-off-by: Elias Probst <mail@eliasprobst.eu>
---
 automx2/server.py       | 13 +++++++++++++
 contrib/automx2.service |  1 +
 requirements.txt        |  1 +
 3 files changed, 15 insertions(+)

diff --git a/automx2/server.py b/automx2/server.py
index a69d991..7caa442 100644
--- a/automx2/server.py
+++ b/automx2/server.py
@@ -16,6 +16,8 @@ GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with automx2. If not, see <https://www.gnu.org/licenses/>.
 """
+import sdnotify
+
 from flask import Flask
 from flask_migrate import Migrate
 from werkzeug.middleware.proxy_fix import ProxyFix
@@ -46,9 +48,12 @@ def _proxy_fix():
 
 
 app = Flask(__name__)
+
+sdnotify.SystemdNotifier().notify("STATUS=Configuring automx2 server")
 app.config['SQLALCHEMY_DATABASE_URI'] = config.db_uri()
 app.config['SQLALCHEMY_ECHO'] = config.db_echo()
 app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
+sdnotify.SystemdNotifier().notify("STATUS=Setting up Flask URL rules")
 app.add_url_rule('/', view_func=SiteRoot.as_view('root'), methods=['GET'])
 app.add_url_rule(APPLE_CONFIG_ROUTE, view_func=mobileconfig.AppleView.as_view('apple'), methods=['GET'])
 app.add_url_rule(INITDB_ROUTE, view_func=InitDatabase.as_view('initdb'), methods=['DELETE', 'GET', 'POST'])
@@ -57,5 +62,13 @@ app.add_url_rule(MSOFT_ALTERNATE_ROUTE, view_func=autodiscover.OutlookView.as_vi
 app.add_url_rule(MSOFT_CONFIG_ROUTE, view_func=autodiscover.OutlookView.as_view('ms1'), methods=['POST'])
 _proxy_fix()
 
+sdnotify.SystemdNotifier().notify("STATUS=Initializing Flask application")
+
 db.init_app(app)
+
+sdnotify.SystemdNotifier().notify("STATUS=Applying DB migrations")
+
 migrate = Migrate(app, db)
+
+sdnotify.SystemdNotifier().notify("STATUS=automx2 server is ready and accepting requests")
+sdnotify.SystemdNotifier().notify("READY=1")
diff --git a/contrib/automx2.service b/contrib/automx2.service
index 5d622e1..13f7b85 100644
--- a/contrib/automx2.service
+++ b/contrib/automx2.service
@@ -10,6 +10,7 @@ ExecStart=/srv/www/automx2/bin/flask run --host=127.0.0.1 --port=4243
 Restart=always
 User=automx2
 WorkingDirectory=/var/lib/automx2
+Type=notify
 
 [Install]
 WantedBy=multi-user.target
diff --git a/requirements.txt b/requirements.txt
index e2b1ac7..4a78fc4 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -5,3 +5,4 @@ SQLAlchemy>=1.3.11
 alembic>=1.3.1
 Flask-Migrate>=2.5.2
 ldap3>=2.6
+sdnotify>=0.3.2
-- 
2.47.0
```

**Comment by rseichter** [comment]
Now I can better understand what motivated you. I am all for supporting Nix (and I switched some of my own machines over to NixOS). Too bad NixOS chose systemd. 😐

What does cause me a bit of unease is the thought of importing a custom Python package like `sdnotify`. It is not available as a Gentoo Linux ebuild yet, so I would break my existing `automx2` ebuild by introducing that additional dependency. At least a sdnotify equivalent should be quite easy to implement; perhaps I'll consider that.

**Comment by eliasp** [comment]
Yes, since `sd_notify()` is a really trivial API, it might make sense to re-implement the required parts just on your own.

Even though I wrote a lot of ebuilds for nearly two decades, it's been a while now since I migrated to NixOS so I don't feel that comfortable anymore in doing so - otherwise I'd just provided an sdnotify ebuild.

**Comment by rseichter** [comment]
Writing an ebuild for sdnotify is not a technical concern per se, but getting any new ebuild accepted by Gentoo certainly is. It has proven to be a significant hassle in the past, and I am not feeling motivated to open that particular can of worms again. I'd rather be migrating more machines to NixOS. I prefer OpenRC to systemd, but that's not enough reason for me to stick with Gentoo.

**Comment by eliasp** [comment]
If you don't want to pull in a mandatory dependency, you could also [make sdnotify support optional](https://github.com/eliasp/automx2/commit/b542621a98a17f3494f3b9d1142cd1d9fd6a7a9e) (same conditions apply to this patch as for the 1st one above).

This would also allow for Gentoo to ship automx2 even without requiring an sdnotify ebuild and later, once it might be part of the tree, to apply the `systemd` USE flag for this optional dependency.

**Comment by rseichter** [comment]
I have published a developer [pre-release](https://pypi.org/project/automx2/#history) with a modernised build configuration and sd_notify support. Perhaps you could take it for a spin?

**Comment by eliasp** [comment]
@rseichter thanks a lot, that's fantastic! Will give it a try ASAP and let you know about my results.

**Comment by rseichter** [comment]
I have successfully tested the pre-release with Arch and updated my Gentoo build. I hope your tests will go well, too.

**Comment by eliasp** [comment]
Same here for me on NixOS - works just as expected - no issues to be seen.

Looking forward to the official release, so I can drop the downstream patch.

**Comment by rseichter** [comment]
The release is published. I only just realised that I did not ask for permission to mention your name in the change log; I hope that's not a problem?

**Comment by eliasp** [comment]
Thanks a ton! Will use the release from here on.
No worries about mentioning me in the ChangeLog - very much appreciated!
